proposals/lip-0052.md

```diff
@@ -362,712 +365,1901 @@ The module provides the following commands to modify the NFT store.
+```
+
 ##### Execution
 
 When executing a cross-chain NFT transfer message `ccm`, the logic below is followed.
 
 ```python
-nftID = ccm.params.nftID
-chainID = nftID.chainID
-sendingChainID = ccm.sendingChainID
-senderAddress = ccm.params.senderAddress
-recipientAddress = ccm.params.recipientAddress
-receivedAttributes = ccm.params.attributes
-ownChainID = interoperability.getOwnChainAccount().ID
-
-if (chainID not in [ownChainID, sendingChainID]
-    or length(senderAddress) != ADDRESS_LENGTH
-    or length(recipientAddress) !=  ADDRESS_LENGTH
-    or length(receivedAttributes) > MAX_BYTE_LENGTH_ATTRIBUTES
-    or (chainID == ownChainID
-        and NFTStore(nftID).owner != uint32(sendingChainID))
-    or (chainID == sendingChainID
-        and NFTStore(nftID) exists)):  
-    if  (ccm.fee >= MIN_RETURN_FEE * length(ccm)
-        and ccm.status == CCM_STATUS_OK):
-        interoperability.error(ccm, CCM_STATUS_NFT_PROTOCOL_VIOLATION)
-    interoperability.terminateChain(sendingChainID)
-    stop ccm execution
-
-if chainID == ownChainID:
-    oldAttributes = NFTStore(nftID).attributes
-    if ccm.status == CCM_STATUS_OK:
-        newAttributes = getNewAttributes(nftID.collection,
-                                         oldAttributes,
-                                         receivedAttributes)
-        newRecipientAddress = recipientAddress
-    else:
-        newAttributes = oldAttributes
-        newRecipientAddress = senderAddress
-
-    nftID = canonicalNFTID(nftID)
-    NFTStore(nftID).owner = newRecipienAddress
-    NFTStore(nftID).attributes = newAttributes
-    createUserEntry(newRecipientAddress, nftID)
-
-else: # chainID == sendingChainID
-    if NFTSupported(nftID) == FALSE:
-        if (ccm.fee >= MIN_RETURN_FEE * length(ccm)
-            and ccm.status == CCM_STATUS_OK):
-            interoperability.error(ccm, CCM_STATUS_NFT_NOT_SUPPORTED)
-        stop ccm execution
-    if ccm.status == CCM_STATUS_OK:
-        createNFTEntry(nftID,
-                       recipientAddress,
-                       receivedAttributes)
-        createUserEntry(recipientAddress, nftID)
-    else:
-        createNFTEntry(nftID,
-                       senderAddress,
-                       receivedAttributes)
-        createUserEntry(senderAddress, nftID)
+def execute(
+    ccu: Transaction,
+    ccm: CCM
+) -> None:
+
+    nftID = ccm.params.nftID
+    senderAddress = ccm.params.senderAddress
+    recipientAddress = ccm.params.recipientAddress
+    receivedAttributes = ccm.params.attributes
+    data = ccm.params.data
+    receivingChainID = ccm.receivingChainID
+    sendingChainID = ccm.sendingChainID
+
+    if getChainID(nftID) == OWN_CHAIN_ID: # execution on the native chain
+        storedAttributes = NFTStore(nftID).attributesArray
+        if ccm.status == CCM_STATUS_OK:
+            NFTStore(nftID).owner = recipientAddress
+            NFTStore(nftID).attributesArray = getNewAttributes(nftID, storedAttributes, receivedAttributes)
```

Comment:
This logic of CCM processing will not allow modules to define custom handling of the NFT attributes. It will be just centrally decided from the NFT module (in particular, by the `getNewAttributes` function).

I think, the following could be a solution: every module that works with NFT attributes should define `afterCrossChainCommandExecution` CCM hook (see https://github.com/LiskHQ/lips/blob/main/proposals/lip-0049.md#apply). In this hook, the module will have access to the CCM with the attributes from another chain and `nftID`, to the `getAttributes` and `setAttributes` methods of the NFT module. Then they can implement any custom logic, for example update some attributes and discard some other ones.

Note that the default behavior for all modules that have not implemented `afterCrossChainCommandExecution` will be correct: the received attributes will be discarded. 

I am also interested to hear from @ricott1 about this.

Reply:
Noted!